theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
```

Comment:
The mapping will contain the mapping for all Theano variable in the graph.
We won't put 2 times the same variable in the mapping (I think this is your question) for a simple reason, we only add them to the mapping when they are produced (so input of the graph and the outputs of apply node).

If we where adding them to the mapping when we would use them, then yes, there would be case where the variable would already be in the graph.